lib/puppet/provider/puppet_certificate/ruby.rb

```diff
@@ -1,127 +1,188 @@
 require 'fileutils'
 require 'puppet/face'
+require 'puppet/ssl/certificate'
 
 Puppet::Type.type(:puppet_certificate).provide(:ruby) do
-  desc "Manage Puppet certificates using the certificate face"
+  desc 'Manage Puppet certificates'
+
+  def initialize(value = {})
+    super(value)
+    if at_least_puppet6
+      @cert_provider = Puppet::X509::CertProvider.new(hostprivkey: nil, hostcert: nil)
+      @machine = Puppet::SSL::StateMachine.new
+    else
+      debug('using legacy code and APIs')
+    end
+  end
 
   def create
     debug "create #{@resource[:name]}"
-    generate_key
     submit_csr
-    if ca_location == 'local'
-      sign_certificate
-    else
-      retrieve_certificate
-    end
+    sign_certificate if @resource[:ca_location] == :local
+    retrieve_certificate
   end
 
-  def generate_key
-    unless key
-      debug "generating new key for #{@resource[:name]}"
-      ensure_cadir if ca_location == 'local'
-      Puppet::SSL::Oids.register_puppet_oids
+  def submit_csr
+    debug "generating new key for #{@resource[:name]} and submitting csr"
+    ensure_cadir if @resource[:ca_location] == :local
+    Puppet::SSL::Oids.register_puppet_oids if Puppet::SSL::Oids.respond_to?('register_puppet_oids')
+
+    if at_least_puppet6
+      key = OpenSSL::PKey::RSA.new(Puppet[:keylength].to_i)
+      @cert_provider.save_private_key(@resource[:name], key)
+
+      csr = create_request(key)
+      ssl_context = @machine.ensure_ca_certificates
+      route = create_route(ssl_context)
+      begin
+        route.put_certificate_request(@resource[:name], csr, ssl_context: ssl_context)
+      rescue Puppet::HTTP::ResponseError => e
+        if e.response.code == 400
+          raise Puppet::Error.new(_("Could not submit certificate request for '%{name}' to %{url} due to a conflict on the server") % { name: @resource[:name], url: route.url }, e)
+        end
+
+        raise Puppet::Error.new(_('Failed to submit certificate request: %{message}') % { message: e.message  }, e)
+      rescue StandardError => e
+        raise Puppet::Error.new(_('Failed to submit certificate request: %{message}') % { message: e.message  }, e)
+      end
+      @cert_provider.save_request(@resource[:name], csr)
+      Puppet.notice(_("Submitted certificate request for '%{name}' to %{url}") % { name: @resource[:name], url: route.url })
+    else
       host = Puppet::SSL::Host.new(@resource[:name])
-      host.generate_certificate_request(:dns_alt_names => options[:dns_alt_names])
+      dns_alt_names = @resource[:dns_alt_names].join(',') if @resource[:dns_alt_names]
+      host.generate_certificate_request(:dns_alt_names => dns_alt_names)
```

Comment:
What is the value of `dns_alt_names`, if `@resource[:dns_alt_names]` is nil? Can that happen?

Reply:
Also `nil` and I think that's fine and isn't new.  All the same, I can take a look to make this clearer.

Reply:
Fixed